ethcore/src/engines/mod.rs

```diff
@@ -338,6 +340,27 @@ pub trait Engine<M: Machine>: Sync + Send {
 	fn is_timestamp_valid(&self, header_timestamp: u64, parent_timestamp: u64) -> bool {
 		header_timestamp > parent_timestamp
 	}
+
+	/// Generate metadata for a new block.
+	fn generate_metadata(&self, bytes: &[u8], provider: &M::BlockProvider) -> M::BlockMetadata;
```

Comment:
I'm wary of providing the entire blockchain context here. We do it in a specialized way through `Ancestry` in some localized points. In most cases we won't need anything more than the `parent_metadata`.

I would recommend changing the signature to `generate_metadata(header, ancestry)`. Then we can invoke the rolling finality tracker in `authority_round` from there and embed the most recently-finalized block hash in there and use it in the fork choice. That will pose a problem with strongly-typed metadata according to the machine because if we've defined the "machine" metadata to be only a `TotalDifficulty` then that's not enough for that modification `authority_round`'s fork choice.

The other option is to avoid strongly-typed metadata entirely. This is kind of necessary until we stop using trait-object engines.